doc/adr/0009-container-image-tags.md

```diff
@@ -0,0 +1,55 @@
+# 9. Convention for tagging of container images
+
+Date: 2020-03-13
+
+## Status
+
+Proposed
+
+## Context
+
+Libero's products are assembled from product specific, general libero and external container images.
+
+Integration testing and reproducible deploys require uniquely identifiable image tags.
+
+
+## Decision
+
+### Tagging flow
+
+- each container image of a Libero container shall be tagged with the git-sha of the build
+- compose files used for integration test shall use immutable image tags
+- the compose files will track the component images and external images with the help of PR generating bots (e.g. dependabot, renovate)
+- upon passed integration tests the used libero component images get tagged with the git-sha of the integration/umbrella repo
+- semver tagging of Libero component images is triggered by tagging of the umbrelle repo
```

Comment:
This is concerning. Components _have_ to be versioned independently; this smells of a distributed monolith.

Reply:
But isn't this the beauty of tagging? One image can have multiple tags. One that refers to the component version and one the the umbrella version. To be fair my view is probably skewed as coming from someone who wants to understand the deploy and not the components.

Reply:
Umbrella version being an eLife deployment? Given this is Libero, _all_ that knowledge of what eLife is deployment has to be on eLife's side.

Reply:
To take a very complicated example from the Debian/Ubuntu world:
```
 apt-cache madison vim
       vim | 2:7.4.1689-3ubuntu1.3 | http://gb.archive.ubuntu.com/ubuntu xenial-updates/main amd64 Packages
```
`2` is an optional [epoch](http://www.fifi.org/doc/debian-policy/policy.html/ch-versions.html) and can be ignored here.
`7.4.1689` is the component version, comes from upstream and is what you see with `--version`.
`3ubuntu1.3` is the distribution's package version (it it was Debian, it would just be `0`, `1`, `2`, ...), however this is not consistent across packages as it advances independently in each of them.

Reply:
Since the umbrella repository is not as complex as a Linux distribution, it would publish images after integration testing; so the `3ubuntu1.3` part could be consistent for each of them, but not at the expense of individual versioning of components.

Reply:
What do you mean by umbrella repository?

Reply:
Umbrella would be the repo that combines various components into the Editor, Publisher, Reviewer product. To be honest I had only studied the Reviewer in depth.

So umbrella repo is the place for integration/end2end test to reside. This requires a compose file and/or helm chart in that repo.

Reply:
> all that knowledge of what eLife is deploying has to be on eLife's side.

This could probably be added to `Context`?

From my point of view, I don't see an umbrella version (if it exists) as an eLife-specific concept, but as a product version. Given installation X (eLife, a PRC platform, an IJM project, etc) choosing a single version of Libero Reviewer or Publisher it's easier than figuring out which set of images work together. This product version may have some correlation with the Helm chart version where one exists.

The alternative is that only component-specific versions are in place, but it puts the onus on the user to find out what works together unless strong constraints are put on these versions (e.g. `1.1.x` of component A works with `1.1.y` of component B) which is monolithic.

Reply:
https://github.com/libero/publisher-chart seems to be the umbrella repo for Publisher. It wasn't very clear what [version number](https://github.com/libero/publisher-chart/blob/master/helm/publisher/Chart.yaml#L7) to give to it.

Reply:
> a product version

This is the crux of it. First time this has been mentioned. In https://github.com/libero/community/blob/master/doc/adr/0008-libero-infrastructure.md we're not supporting using higher-levels that group components together.

> single version of Libero Reviewer or Publisher
> It wasn't very clear what version number to give to it.

Exactly, I've no idea what a version of 'Publisher' is given it's a pick-and-choose-and-extend-and-replace setup...

> what works together 

Unless it's a major version bump, any other changes means semver is not followed. (Though bugs can break things of course.)